### ISSUE TITLE
Portal: Investigate FE feature parity between the current reader implementation and the new design

### ISSUE THREAD

**Issue by mashikag** [Automattic/wp-calypso#73937]
We may take the current implementation of the Reader's followed sites manager and work of it by incrementally making it adaptable for external users subscription management and then applying the new design to it. 

However first we need to investigate the feature parity between what the current implementation offers and what we need to offer, as well as estimating the complexity of making the component API agnostic (external users API vs. internal WPCOM users API).

**Comment by mashikag** [comment]
## High level UI investigation
Currently the view at [https://wordpress.com/following/manage](https://wordpress.com/following/manage) allows wpcom user to manage their subscriptions. It displays a list of subscribed-to sites and allows for certain actions on those sites.

### Current Impl. in Reader VS. New Design for External Users
TLDR; No difference between the two in terms data displayed and actions.

Comparing the page to the new design for subscription management for external users does not expose any differences in terms of the subscription actions a user can take. 

The “View: All” select input is an exception (marked with encircled question mark in the new design screenshot). For now I am not entirely sure what it is meant to do so I will not include it in my comparison.

![Image](https://user-images.githubusercontent.com/2019970/222157046-eede7ac9-ade0-4eee-b69c-3f2cbce02ac9.png)

Subscribed sites actions include:
1. Search by site name or address
2. Sort by
3. Go to the site
4. Toggle “Notify me of new posts”
5. Toggle “Email me new posts”
6. Modify frequency of new posts emails
7. Toggle “Email me new comments”
8. Unfollow

### Current Impl. in Reader VS. New Design for WPCOM Users
However comparing the current implementation against the new design for wpcom (internal) user, seems to introduce new actions and data to be displayed.

![Image](https://user-images.githubusercontent.com/2019970/222157299-54961d41-8fb6-4940-8dec-067209e6330c.png)



New actions:
1. Manage purchases

New data to be displayed:
1. “Paid” tag
2. “P2” tag

### Conclusion
It still looks like it may be a viable option to make the current implementation of the subscribed site's table work for the new subscription management portal for the external users', and to start applying the new design to it from there.

#### Next steps
Investigate the dev. complexity of making the current subscribed sites table component applicable to the new portal for the external users.

@yansern

**Comment by mashikag** [comment]
# Code Deep Dive

## Components

### Site row component

[The `siteRowRenderer` function](https://github.com/Automattic/wp-calypso/blob/60a0b2727efecb1c20ed5e7bcc84f7a28c3753d5/client/reader/components/reader-infinite-stream/row-renderers.js) is used to render the subscribed-to site row in the current implementation of subscription management for internal (wpcom) users. The function is essentially based on the [`ReaderSubscriptionListItem` component](https://github.com/Automattic/wp-calypso/blob/27517d64f25fb62c0424e78d58081a787e077021/client/blocks/reader-subscription-list-item/index.jsx#L26).

The are another two different components that utilize the shared [`siteRowRenderer`](https://github.com/Automattic/wp-calypso/blob/60a0b2727efecb1c20ed5e7bcc84f7a28c3753d5/client/reader/components/reader-infinite-stream/row-renderers.js). One of them is on the same subscription management page and another one is on Reader's search page:

1. [`FollowingManageSubscriptions` component](https://github.com/Automattic/wp-calypso/blob/27517d64f25fb62c0424e78d58081a787e077021/client/reader/following-manage/subscriptions.jsx)
   - It uses the component to render user's subscribed to sites.
   - It allows a user to manage notifications for the sites and to unfollow the site
   - Found at https://wordpress.com/following/manage
![Screen Shot 2023-03-01 at 16 45 04](https://user-images.githubusercontent.com/2019970/222376049-9efe124b-ca76-4682-aaab-b01d581b6b09.png)

2. [`FollowingManageSearchFeedsResults` component](https://github.com/Automattic/wp-calypso/blob/85e12ee6f4729343555ad78d6e90f4a7f237a0f4/client/reader/following-manage/feed-search-results.jsx)
    - It uses the component to display user's feed search results at the Reader's **Following -> Manage** page
    - Found at https://wordpress.com/following/manage
  ![Screen Shot 2023-03-02 at 09 34 01](https://user-images.githubusercontent.com/2019970/222375654-723dc33f-6044-4d8e-b5bc-89cb813e86b2.png)

3. [`SiteResults` component](https://github.com/Automattic/wp-calypso/blob/f2e6f88ae65c8571bbac21b284c29b48ec987598/client/reader/search-stream/site-results.jsx)
   - It uses the component to display wpcom sites result after user makes a search query at the Reader's "Search" page
   - Found at https://wordpress.com/read/search
![Screen Shot 2023-03-01 at 16 44 36](https://user-images.githubusercontent.com/2019970/222376336-978accab-8cb5-4594-b683-ceeee87cd443.png)

Hence the [`ReaderSubscriptionListItem` component](https://github.com/Automattic/wp-calypso/blob/27517d64f25fb62c0424e78d58081a787e077021/client/blocks/reader-subscription-list-item/index.jsx#L26) supports quite few variations, e.g.: no **Settings** button, follow site button instead of unfollow site button etc. 

### Subscription settings component
[`ReaderSiteNotificationSettings` component](https://github.com/Automattic/wp-calypso/blob/31e64d93932ce9287e7e2e3702249c1e52fa3bc2/client/blocks/reader-site-notification-settings/index.jsx) used in [the `ReaderSubscriptionListItem` component](https://github.com/Automattic/wp-calypso/blob/27517d64f25fb62c0424e78d58081a787e077021/client/blocks/reader-subscription-list-item/index.jsx#L167). 
![Screen Shot 2023-03-02 at 11 47 28](https://user-images.githubusercontent.com/2019970/222407487-98c789aa-73b8-45a6-9f06-31ba0bf5eac4.png)

This settings component is also shared with other reader components. We can find it in the following places.

1. In the Reader in a selected site's feed
- Example URL: https://wordpress.com/read/feeds/138765036
![Screen Shot 2023-03-02 at 12 42 55](https://user-images.githubusercontent.com/2019970/222419788-53f0c74d-befb-4a5e-9fdf-ac62b9bf259a.png)

2. In the Reader's "site-stream". However I do not know how to get to this screen other than using the direct URL.
 - Example URL: https://wordpress.com/read/blogs/115523683
 - Almost identical to "feed-stream" from point 1 above
![Screen Shot 2023-03-02 at 13 01 23](https://user-images.githubusercontent.com/2019970/222423976-8e95d884-a967-46a8-9e85-904081d4f3ef.png)


## Challenges
Below I am listing some of the challenges that would come if we were to build on top of the current implementation.

### Ensuring no regressions
It may be an unnecessary overhead for us to constantly ensure that we are not breaking the implementation in other places of the Reader (see the **Site row component** above for more details).

### Class components
All the reader components involved in rendering the sites list are class based components. General consensus is to move away from classes and to build functional components instead. 

### Client state
All the reader related data is stored within redux store that is defined within the `client` package inside `client/state`. The state is not available to outside packages. 

## Conclusions
I think it could end up being quite costly if we were to decide to build up on top of the current reader implementation. 
I think we should consider building our own light-weight site row components instead. 

## Next steps
The following are not the next steps to this task, but rather the remaining steps that we should take on as part of dev planning:
- Identify the exact components we need to build
  -  Keep in mind to make them easily integrable for other parts of the reader
- Decide on the data source for our new portal:
   - Are we going to depend on the calypso's redux store?
     -  if so - consider that the store is not accessible to outside packages and we will have to inflate our package components from the `client` package
     - Consider switching to use instead:
       - [`packages/data-stores`](https://github.com/Automattic/wp-calypso/tree/trunk/packages/data-stores)
       - [React Query](https://react-query-v3.tanstack.com/)

**Comment by yansern** [comment]
Thanks for the deep dive Maciej!

> I think we should consider building our own light-weight site row components instead.

I'd like you to also consider other potential work that may be involved around building our own lightweight site row components, as what may be conceived as lightweight initially could potentially inflate over time, with things like lazy loading, sorting, searching, empty & error states, and (possibly) e2e tests.

What I also noticed in the current Reader implementation is how every single action (follow/unfollow/update frequency/toggling every setting) has tracks recorded, even the act of scrolling the lazy-loaded list up and down has a tracks event recorded.

> General consensus is to move away from classes and to build functional components instead.
> I think it could end up being quite costly if we were to decide to build up on top of the current reader implementation.

My biggest concern is whether it will be arguably more costly not to build on top of the current reader implementation as those other things as mentioned above already exists in Reader.

Is there a middle ground that we can take? For example, improve upon the existing component with certain things allowed:
- OK to migrate or duplicate individual components from class components to function components as we go / as needed.
- OK to decouple data from components and extract those individual calypso components into packages as we go / as needed.

> Are we going to depend on calypso's redux store?

Could we have an assessment of what the existing Reader data store already offers and how it is built?

This is because, on the Reader API end, we'll just be riding on the existing API infrastructure, we just need to add Cookie Auth to support external users (email-logged-in-non-WPCOM-users). The existing data store will likely continue to work with zero or minimal modification.

> if so - consider that the store is not accessible to outside packages and we will have to inflate our package components from the client package

By "inflate", you mean, Calypso components that retrieves the data from the Calypso data store and then feeds the data into the package components which are data agnostic, right?

> Consider switching to use instead:
>  - [packages/data-stores](https://github.com/Automattic/wp-calypso/tree/trunk/packages/data-stores)
>  - [React Query](https://react-query-v3.tanstack.com/)

Migrating from the Reader data store from Calypso to data store packages can be an easy task IF there are no other Calypso components that depend on the Reader data store.

It is easier to estimate if this is doable after performing an assessment on the existing Reader data store (as mentioned/requested above).

As for React Query, I'm not sure if this would be too much of a paradigm shift, but I'd be happy to be proven wrong.

**Comment by mashikag** [comment]
> What I also noticed in the current Reader implementation is how every single action (follow/unfollow/update frequency/toggling every setting) has tracks recorded, even the act of scrolling the lazy-loaded list up and down has a tracks event recorded.

True. Let's ensure that we include tracking in the acceptance criteria of the relevant tasks. 


> Is there a middle ground that we can take? For example, improve upon the existing component with certain things allowed:
> 
> - OK to migrate or duplicate individual components from class components to function components as we go / as needed.
> - OK to decouple data from components and extract those individual calypso components into packages as we go / as needed.

I think this is essentially what we will do. Duplicate individual components and:
1. Migrate them from class components to functional components.
2. Decouple data, i.e. make components data source agnostic.

> By "inflate", you mean, Calypso components that retrieves the data from the Calypso data store and then feeds the data into the package components which are data agnostic, right?

Correct.

---
> Could we have an assessment of what the existing Reader data store already offers and how it is built?

Yes, I created another task where I will handle it: https://github.com/Automattic/wp-calypso/issues/74079. Let's continue all the data related discussions there. 

I will close this task as completed. However if you feel like some questions still remain unanswered let's re-open it and work them out.
